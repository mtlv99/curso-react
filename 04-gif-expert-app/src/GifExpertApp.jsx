import React, { useState } from 'react';

// Los archivos de index/indice o de barril son utiles para este tipo de importaciones,
// 2 componentes en un mismo import
import { AddCategory, GifGrid } from './components';

const GifExpertApp = () => {
  const [category, setCategory] = useState([]);

  const onAddCategory = (newCategory) => {
    // Valida que no exista una categoria con el mismo nombre, para evitar el error de las keys en los map.
    // No es mejor usar un uuid?
    if (category.includes(newCategory)) return;

    // Nota:
    // (A partir de React 18, si hay varios setState seguidos, React esperará a que ésta funcion principal
    // terminar para hacer el re-renderizado. En versiones anteriores, se hace un rerender por cada setState )

    setCategory([newCategory, ...category]);
    // setCategory((categories) => [newCategory, ...categories]);
    // Los llamados de arriba son equivalentes.
    // Aunque en componentes hijos, NO es necesario pasar el state de 'category' por props! Solo usar el callback de setCategory.
  };

  return (
    <>
      <h1>Buscador de GIF&apos;s</h1>

      <AddCategory
        // category={category}
        // setCategory={setCategory}
        // Recordar que cualquier funcion que tenga un 'on' al principio,
        // indica que emite un valor hacia afuera del componente (usualmente).
        onNewCategory={onAddCategory}
      />

      {category.map((categoryItem) => (
        <GifGrid
          key={categoryItem}
          category={categoryItem}
        />
      ))}
    </>


  );
};

export default GifExpertApp;
