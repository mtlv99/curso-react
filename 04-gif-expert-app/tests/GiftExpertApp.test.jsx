import { render, screen } from '@testing-library/react';
import GifExpertApp from '../src/GifExpertApp';

// TODO: hacer mas pruebas
describe('Pruebas en <GifExpertApp />', () => {
  test('Debe de hacer match con el snapshot', () => {
    const { container } = render(<GifExpertApp />);
    expect(container).toMatchSnapshot();
    // screen.debug();
  });
});
