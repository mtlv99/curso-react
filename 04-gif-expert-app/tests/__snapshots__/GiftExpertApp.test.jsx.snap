// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifExpertApp /> Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>
    Buscador de GIF's
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Escribe algo interesante (cafÃ©, galletas, navidad)..."
      type="text"
      value=""
    />
  </form>
</div>
`;
